### ISSUE TITLE
buggy local.getlocale() affects time module (Windows 11)

### ISSUE THREAD

**Issue by Jean-Luc-Picard-2021** [pypy/pypy#5034]
Not sure whether this has to do that I suddently don't
get de_CH anymore but German_Switzerland, or that UTF8
terminal detection is not yet working. With recent CPython I get:
```
>D:\Software2\Logic\Python313\python.exe
Python 3.13.0b4 (tags/v3.13.0b4:567c38b, Jul 18 2024, 10:14:53) [MSC v.1940 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import locale
>>> locale.getlocale()
('German_Switzerland', 'utf8')
>>>
```
With recent PyPy I get:
```
>D:\Software2\PHP\pypy3.10-v7.3.17-win64\python.exe
Python 3.10.14 (39dc8d3c85a7, Aug 27 2024, 14:33:33)
[PyPy 7.3.17 with MSC v.1929 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>>> import locale
>>>> locale.getlocale()
(None, None)
```
I was rather expecting that they agree. Is there some workaround?

**Edit 2024.09.24:** 
This impacts for example the Python time module:
```
/* CPython */
>>> locale.setlocale(locale.LC_TIME, locale.getlocale())
'German_Switzerland.utf8'
>>> time.strftime('%a', time.localtime())
'Di'

/* PyPy */
>>>> locale.setlocale(locale.LC_TIME, locale.getlocale())
'C'
>>>> time.strftime('%a', time.localtime())
'Tue'
```
The time formatting doesn't agree anymore.

**Comment by Jean-Luc-Picard-2021** [comment]
Ok, this is also intefesting, again the same two versions:
```
/* CPython */
>>> locale.getdefaultlocale()
<python-input-2>:1: DeprecationWarning: 'locale.getdefaultlocale' is deprecated and slated for removal in Python 3.15. Use setlocale(), getencoding() and getlocale() instead.
  locale.getdefaultlocale()
('de_CH', 'cp65001')

/* PyPy */
>>>> locale.getdefaultlocale()
('', 'utf-8')
```

**Comment by mattip** [comment]
> I suddently don't get `de_CH` anymore

Did you previously get that? What are the implications of that inconsistency with CPython? 

I have been doing some work on encoding/decoding on windows on py3.10, that is only available in the nightlies. So far I think I have cleaned up
- handling of filenames, which fixes `os ` interfaces like `os.mkdir` and `os.listdir` for files and directories with non-ascii names.
- running PyPy inside a directory with a non-ascii name

I still need to fix
- running `pypy <non-ascii-filename>.py`
- there may be some problems with `pypy -m` or `pypy -c` and non-asii, testing needed

The problem with the console output showing garbage is a little harder to fix. There is a WIP implementation of the PEP 528 `_io._WindowsConsoleIO`, but when stress-testing it, it crashes. See #2999

**Comment by Jean-Luc-Picard-2021** [comment]
If I am not mistaken, in the beginning I always got `de_CH`.
That it now gives `German_Switzerland` repespectivaly '',
has two impacts:

- **Impact 1:**  time library
   Thats my major concern now. The test case shows that
   in CPython mostlikely it takes `German_Switzerland` and nevertheless
   works correctly, and in PyPy it mostlikely takes '' and doesn't work correctly.

- **Impact 2:** localization in my application
   Yes because `de_CH` changed to `German_Switzerland`, I have 
   some trouble in my application. But I could internally map
   `German_Switzerland` to `de_CH` somehow. 

Impact 2 is not in the scope of this ticket. I only raised the ticket
because of Impact 1, i.e. impact on time library.

**Comment by mattip** [comment]
Thanks. I think that is actionable, we have a [test](https://github.com/pypy/pypy/blob/ca0472fe02069a4795c565d657f6bd8e6e1de8e2/pypy/module/time/test/test_time.py#L511) for setting locale and calling time, but it is currently skipped since I did not realize it was important. I will mark this for the next release.

**Comment by Jean-Luc-Picard-2021** [comment]
Maybe its an issue how I install PyPy? In just manually 
unzip this download and use the python.exe there:
```
pypy3.10-v7.3.17-win64.zip
```
Should I change some settings as well. CPython is
installed differently on my side. For CPython I leave
it to an installer to install it.

**Comment by mattip** [comment]
Unzipping PyPy is the standard way to install it.

For now, a work-around would be to call `locale.setlocale(locale.LC_TIME, ('German_Switzerland', 'utf8'))` somewhere in your program before using the data functions. I will try to figure out where our startup code is not properly setting the locale.

**Comment by Jean-Luc-Picard-2021** [comment]
Was hoping for a command line option. For example
in Java I can use the following JVM options:
```
-Duser.language=de -Duser.region=CH
```
Which overrides any automatic detection before launching
the given main class. But I am new to Python, so I don't know 

whether such command line options do exist.

**Comment by mattip** [comment]
Some notes for future me:

`locale.getlocale()` calls `_locale.setlocale(locale.LC_CTYPE)`

On CPython3.10 (on my machine),  this returns
```
LC_CTYPE=English_United States.1252
```

On PyPy3.10, it returns only `C`. This result is then fed to `locale._parse_localname(localname)`, which ends up returning `(None, None)` on PyPy. So the difference is somewhere in `_locale.setlocale()` in RPython

**Comment by Jean-Luc-Picard-2021** [comment]
Does RPython have a separate issue tracker?
If yes, maybe move the ticket there. I don't know,

I am new to RPython, this is always bundled with 
the PyPy JIT-ing compiler, or also standalone available?

**Comment by cfbolz** [comment]
Rpython is completely contained in the PyPy repo, there's no separate issue tracker. So this is the right place :-)

**Comment by mattip** [comment]
The magic happens [here](https://github.com/python/cpython/blob/0c5fc27217525c4e40b4064e6979f467540c2fc8/Python/pylifecycle.c#L436). Before this call to `setlocale(2, "")`, a call to `setlocale(2, NULL)` returns `'C'`, after it the same call returns (on my machine) ` English_United States.1252`. So I guess we need to force that call in our `app_main.py`.

**Comment by mattip** [comment]
The MSVC [documentation](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/setlocale-wsetlocale?view=msvc-170) for `setlocale( LC_ALL, "" );` says
> Sets the locale to the default, which is the user-default ANSI code page obtained from the operating system. The locale name is set to the value returned by [GetUserDefaultLocaleName](https://learn.microsoft.com/en-us/windows/win32/api/winnls/nf-winnls-getuserdefaultlocalename). The code page is set to the value returned by [GetACP](https://learn.microsoft.com/en-us/windows/win32/api/winnls/nf-winnls-getacp).

The linux manpage says
> If  locale  is an empty string, "", each part of the locale that should be modified is set according to the environment variables.  The details are implementation-dependent.  For glibc, first (regardless of category), the environment variable LC_ALL is inspected, next the environment variable with the same name  as the category (see the table above), and finally the environment variable LANG.  The first existing environment variable is used.  If its value is not a valid locale specification, the locale is unchanged, and setlocale() returns NULL.

So I guess we should do this on all OSes.

**Comment by Jean-Luc-Picard-2021** [comment]
I just find that the Python documentation says:

> Applications typically start with a call of
> ```
> import locale
> locale.setlocale(locale.LC_ALL, '')
> ```
> This sets the locale for all categories to the user’s default setting

https://docs.python.org/3/library/locale.html

But it seems that executing a Python script or running 
the REPL via python.exe has also this effect.

**Comment by mattip** [comment]
Closed by #5060. Please try a [nightly](https://buildbot.pypy.org/nightly/py3.10/) that includes this fix (i.e. after tonight's builds).

**Comment by Jean-Luc-Picard-2021** [comment]
I tried on Unix recently, with an old pypy3. 
```
Python 3.9.18 (7.3.15+dfsg-1build3, Apr 01 2024, 03:12:48)
[PyPy 7.3.15 with GCC 13.2.0] on linux
```
When I do a complete locale change:
```
$ mkdir -p ~/.locale
$ I18NPATH=./wrk/ localedef -f UTF-8 -i de_CH ~/.locale/de_CH.UTF-8
$ LOCPATH=~/.locale LC_ALL=de_CH.UTF-8 date
$ echo "export LOCPATH=~/.locale" >> ~/.bashrc
$ echo "export LANG=de_CH.UTF-8" >> ~/.bashrc
```
It works all fine:
```
$ pypy3 playerpy/canned/dogelog.py
Dogelog Spieler, Prolog zum Mond, 1.2.4 (20.10.2024)
(c) 1985-2024, XLOG Technologies AG, Schweiz
?- current_prolog_flag(sys_locale, X).
X = de_CH.
```
Will test Windows 11 in a few months when a new pypy3 release is out.
